content/blog/2016/2016-08-25-sauce-pipeline.adoc

```diff
@@ -0,0 +1,242 @@
+The transition to pipeline couldn't be smoother!
+
+image::/images/post-images/2016-08-25/multibranch-report.png[Pipeline Report for "sauce-pipeline" Branch, role="center"]
+
+
+=== Adding Platforms
+Next I added some more platforms to my matrix.
+This is a perfect example of the power of pipeline as code.
+Adding new platforms required changing both the pipeline and the Nightwatch.js configuration file.
+Since my pipeline was now checked in, I could change both as part of one commit and be assured my pipeline and test framework wouldn't get out of sync.
+
+I added three new environments to nightwatch.json:
+
+[source,js]
+----
+"test_settings" : {
+  "default": { /*----8<----8<----8<----*/ },
+  "chrome": { /*----8<----8<----8<----*/ },
+
+  "firefox": {
+    "desiredCapabilities": {
+      "platform": "linux",
+      "browserName": "firefox",
+      "version": "latest"
+    }
+  },
+  "ie": {
+    "desiredCapabilities": {
+      "platform": "Windows 10",
+      "browserName": "internet explorer",
+      "version": "latest"
+    }
+  },
+  "edge": {
+    "desiredCapabilities": {
+      "platform": "Windows 10",
+      "browserName": "MicrosoftEdge",
+      "version": "latest"
+    }
+  }
+}
+----
+
+And I modified my Jenkinsfile to call them:
+
+[source,groovy]
+----
+//----8<----8<----8<----8<----8<----8<----
+try {
+    def configs = [ // <1>
+        'chrome',
+        'firefox',
+        'ie',
+        'edge'
+    ].join(',')
+    // Nightwatch.js support color ouput, so wrap that step for ansi color
+    wrap([$class: 'AnsiColorBuildWrapper', 'colorMapName': 'XTerm']) {
+        // Run selenium tests using Nightwatch.js
+        sh "./node_modules/.bin/nightwatch -e ${configs} --test tests/guineaPig.js" // <2>
+    }
+} //----8<----8<----8<----8<----8<----8<----
+----
+<1> Using an array to improve readability and make it easy to add more platforms later.
+<2> Change from single-quoted string to double-quoted to support variable substitution.
+
+Then I kicked off a new build with these changes and it worked perfectly! ... Almost.
+
+=== Even Test Frameworks Have Bugs
+
+First, the good news.  As expected, Jenkins picked up the new pipeline and ran Nightwatch.js on four platforms.
+Nightwatch.js was already configured to used multiple worker threads to run tests against those platforms in parallel and
+my Sauce Labs account supported enought to run them all at the same time.
+Sauce Labs recorded the results and correctly linked them into the report for this build.
+image::/images/post-images/2016-08-25/platforms-console.png[Nightwatch.js Console Output for Multiple Platforms, role="center"]
+
+image::/images/post-images/2016-08-25/platforms-stages.png[Sauce Lab Parallel Execution, role="center"]
+
+image::/images/post-images/2016-08-25/platforms-results.png[Sauce Labs Results List, role="center"]
+
+The bad news was that JUnit didn't see the results additional platforms.  It still only showed 5 results instead of 20.
```

Comment:
Unclear what the correct wording should be, but "results additional platforms" is confusing to me.